cpp/include/ray/api/logging.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#if defined(_WIN32)
+#ifndef _WINDOWS_
+#ifndef WIN32_LEAN_AND_MEAN  // Sorry for the inconvenience. Please include any related
+                             // headers you need manually.
+                             // (https://stackoverflow.com/a/8294669)
+#define WIN32_LEAN_AND_MEAN  // Prevent inclusion of WinSock2.h
+#endif
+#include <Windows.h>  // Force inclusion of WinGDI here to resolve name conflict
+#endif
+#ifdef ERROR  // Should be true unless someone else undef'd it already
+#undef ERROR  // Windows GDI defines this macro; make it a global enum so it doesn't
+              // conflict with our code
+enum { ERROR = 0 };
+#endif
+#endif
+
+#if defined(DEBUG) && DEBUG == 1
+// Bazel defines the DEBUG macro for historical reasons:
+// https://github.com/bazelbuild/bazel/issues/3513#issuecomment-323829248
+// Undefine the DEBUG macro to prevent conflicts with our usage below
+#undef DEBUG
+// Redefine DEBUG as itself to allow any '#ifdef DEBUG' to keep working regardless
+#define DEBUG DEBUG
+#endif
+
+namespace ray {
+namespace api {
+
+enum class CppRayLogLevel {
+  TRACE = -2,
+  DEBUG = -1,
+  INFO = 0,
+  WARNING = 1,
+  ERROR = 2,
+  FATAL = 3
+};
+
+#ifdef RAY_LOG
+#undef RAY_LOG
+#undef RAY_LOG_INTERNAL
+#define RAY_LOG_INTERNAL(level) *CreateCppLog(__FILE__, __LINE__, level)
+#define RAY_LOG(level) \
+  if (IsLevelEnabled(CppRayLogLevel::level)) RAY_LOG_INTERNAL(CppRayLogLevel::level)
+#endif
```

Comment:
We only re-define `RAY_LOG` if `RAY_LOG` is already defined? I think the correct code is:

```
#ifdef RAY_LOG
  #undef RAY_LOG
#endif

#define RAY_LOG
```

Reply:
And please use indents between `#ifdef` and `#endif`

Reply:
clang-format will remove indents.

Reply:
`
#ifdef RAY_LOG
  #undef RAY_LOG
#endif

#define RAY_LOG
`
The code will cause redefine compile error.